edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+		return fmt.Errorf("fail to generate UpdateDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.UpdateDevice(ctx, udr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcCreateDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	cdmr, err := dmiclient.CreateDeviceModelRequest(model)
+	if err != nil {
+		return fmt.Errorf("fail to create CreateDeviceModelRequest for device model %s with err: %v", model.Name, err)
+	}
+	_, err = dc.CreateDeviceModel(ctx, cdmr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcRemoveDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	rdmr, err := dmiclient.RemoveDeviceModelRequest(model.Name)
+	if err != nil {
+		return fmt.Errorf("fail to create RemoveDeviceModelRequest for device model %s with err: %v", model.Name, err)
+	}
+	_, err = dc.RemoveDeviceModel(ctx, rdmr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcUpdateDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	udmr, err := dmiclient.UpdateDeviceModelRequest(model)
+	if err != nil {
+		return fmt.Errorf("fail to create UpdateDeviceModelRequest for device model %s with err: %v", model.Name, err)
+	}
+	_, err = dc.UpdateDeviceModel(ctx, udmr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetSockPath(protocol string) string {
+	for _, mapper := range mapperList {
+		if mapper.GetProtocol() == protocol {
+			return string(mapper.Address)
+		}
+	}
+	return ""
+}
+
+func initDeviceModelInfoFromDB() {
+	deviceModelList = make(map[string]*v1alpha2.DeviceModel)
+	metas, err := dao.QueryMeta("type", constants.ResourceTypeDeviceModel)
+	if err != nil {
+		klog.Errorf("fail to init device model info from db with err: %v", err)
+		return
+	}
+
+	for _, meta := range *metas {
+		deviceModel := &v1alpha2.DeviceModel{}
+		if err := json.Unmarshal([]byte(meta), &deviceModel); err != nil {
+			klog.Errorf("fail to unmarshal device model info from db with err: %v", err)
+			return
+		}
+		klog.Infof("deviceModel: %+v", deviceModel)
```

Comment:
this is a debug info or we should upper the log level

Reply:
done